kimchi/src/circuits/polynomials/foreign_mul/constraints.rs

```diff
@@ -0,0 +1,40 @@
+//!
+//! Fundamental constraints and related helpers
+//!
+
+use std::cell::RefCell;
+
+use ark_ff::{FftField, One, Zero};
+
+use crate::circuits::expr::{Cache, ConstantExpr, Expr, E};
+
+thread_local! {
+    static CACHE: std::cell::RefCell<Cache>  = RefCell::new(Cache::default());
+}
+
+pub(crate) fn _cache<F: FftField>(mut x: E<F>) -> E<F> {
+    CACHE.with(|cache| x = cache.borrow_mut().cache(x.clone()));
+    x
+}
+
+pub(crate) fn two<F: FftField>() -> E<F> {
+    Expr::Constant(ConstantExpr::Literal(2u32.into()))
```

Comment:
There's already a `From<u64> for Expr<ConstantExpr<F>>`, can you use 2u64.into() below instead?